lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -0,0 +1,949 @@
+  auto proto = C.getProtocol(KnownProtocolKind::CodingKey);
+  TypeLoc protoTypeLoc[1] = { TypeLoc::withoutLoc(proto->getDeclaredType()) };
+  MutableArrayRef<TypeLoc> inherited = C.AllocateCopy(protoTypeLoc);
+
+  auto enumDecl = new (C) EnumDecl(SourceLoc(), C.Id_CodingKeys, SourceLoc(),
+                                   inherited, nullptr, typeDC);
+  enumDecl->setImplicit();
+  enumDecl->setAccessibility(Accessibility::Private);
+
+  auto enumDC = cast<DeclContext>(enumDecl);
+  auto mutableEnumDC = cast<IterableDeclContext>(enumDecl);
+
+  // For classes which inherit from something Codable, we provide case `super`
+  // as the first key (to be used in encoding super).
+  auto classDecl = dyn_cast<ClassDecl>(type);
+  if (classInheritsCodable(tc, classDecl)) {
+    // TODO: Ensure the class doesn't already have or inherit a variable named
+    // "`super`"; otherwise we will generate an invalid enum. In that case,
+    // diagnose and bail.
+    auto super = new (C) EnumElementDecl(SourceLoc(), C.Id_super, TypeLoc(),
+                                         /*HasArgumentType=*/false,
+                                         SourceLoc(), nullptr, enumDC);
+    super->setImplicit();
+    mutableEnumDC->addMember(super);
+  }
+
+  // Each of these vars needs a case in the enum. For each var decl, if the type
+  // conforms to Codable, add it to the enum.
+  auto storedProperties = type->getStoredProperties(/*skipInaccessible=*/true);
+  auto validVarCallback = [&C, &enumDC, &mutableEnumDC](VarDecl *varDecl) {
+    auto elt = new (C) EnumElementDecl(SourceLoc(), varDecl->getName(),
+                                       TypeLoc(), /*HasArgumentType=*/false,
+                                       SourceLoc(), nullptr,
+                                       enumDC);
+    elt->setImplicit();
+    mutableEnumDC->addMember(elt);
+  };
+
+  if (!validateVarsConformToCodable(tc, parentDC, storedProperties,
+                                    validVarCallback))
+    return {nullptr, /* attempted to synthesize? */true};
+
+  // Forcibly derive conformance to CodingKey.
+  tc.checkConformancesInContext(enumDC, mutableEnumDC);
+
+  // "Synthesized" a null enum decl.
+  return {enumDecl, /* attempted to synthesize? */true};
+}
+
+// Creates a new var decl representing
+//
+//   let container : containerBase<keyType>
+//
+// containerBase is the name of the type to use as the base (either
+// KeyedEncodingContainer or KeyedDecodingContainer).
+static VarDecl*
+createKeyedContainer(ASTContext &C, DeclContext *DC, Identifier containerBase,
+                     Type keyType) {
+  // Look up Keyed*Container
+  auto keyedContainerDecl =
+    cast<NominalTypeDecl>(lookupFoundationTypeDecl(C, containerBase));
+
+  // Bind Keyed*Container to Keyed*Container<KeyType>
+  Type boundType[1] = { keyType };
+  auto containerType = BoundGenericType::get(keyedContainerDecl, Type(),
+                                             C.AllocateCopy(boundType));
```

Comment:
`BoundGenericType::get` takes an `ArrayRef` without copying it. Most decls do, in fact, because they likely assume that these things were already allocated earlier in the parsing process (which they were), so it's safe to take a reference without copying. Here, however, I'm synthesizing it all from scratch, so it does need to be allocated because these references are long-lived.